charms/reactive/bus.py

```diff
@@ -443,12 +445,16 @@ def _load_module(root, filepath):
 
 
 def _register_handlers_from_file(root, filepath):
+    exec_whitelist = ('', '.py', '.sh')
     no_exec_blacklist = (
```

Comment:
Instead of a blacklist, would it make sense to check that the first line of the script is some sort of shebang? If it's not, then the shell won't run it anyways.

I see below that you're using `#!/bin/false` in non-scripts, but TBH, I would expect to be able to define a valid handler with that shebang.

Reply:
Technically, the handlers could be any executable, not just a shell script.

Reply:
Check for a shebang or `+x`?

Reply:
https://github.com/juju-solutions/charms.reactive/blob/master/charms/reactive/bus.py#L455

Reply:
if we're already checking for `+x`, then we shouldn't need a blacklist, since nobodies[1] going to mark `copyright` as executable, right?

[1] It might happen accidentally, hence why checking for the shebang, but I think the more important check is `+x`

Reply:
In other words, the logic would be:

 - If `*.py`, then it's a reactive file
 - If it's `+x` and binary, then it's a binary handler
 - If it's `+x` and has a shebang, then it's a script handler
 - If it's `+x` and no shebang, then it's an erroneously-marked `+x` file that should be ignored
 - If it's not `+x` or `*.py`, then it can be ignored

Reply:
Unfortunately, there was an issue (#54) with random files getting marked as executable (I think due to Windows development) which is why the blacklist was added.  While not having the explicit blacklist would be nice, I don't think deciding if a file is binary or not is trivial and TBH, I don't think it's worth putting any more effort into this code since it's pretty deprecated already.  I think the whitelist addition is sufficient to exclude most files that would accidentally get picked up and the blacklist is good enough for the remaining common cases.